c/cardano_lock.c

```diff
@@ -1,18 +1,346 @@
+    if (ret == CKB_INDEX_OUT_OF_BOUND) {
+      break;
+    }
+    if (ret != CKB_SUCCESS) {
+      return ERROR_IDENTITY_SYSCALL;
+    }
+    i += 1;
+  }
+
+  blake2b_final(&blake2b_ctx, msg, BLAKE2B_BLOCK_SIZE);
+
+  return 0;
+}
+
+size_t generate_new_msg(uint8_t *output, size_t output_len,
+                        const uint8_t *payload, size_t payload_len,
+                        const uint8_t *external_aad, size_t external_aad_len) {
+  nanocbor_encoder_t enc;
+  nanocbor_encoder_init(&enc, output, output_len);
+  nanocbor_fmt_array(&enc, 4);
+
+  const char *msg_sign_context = "Signature1";
+  nanocbor_put_tstr(&enc, msg_sign_context);
+  nanocbor_put_bstr(&enc, NULL, 0);
+  nanocbor_put_bstr(&enc, external_aad, external_aad_len);
+  nanocbor_put_bstr(&enc, payload, payload_len);
+
+  return nanocbor_encoded_len(&enc);
+}
+
+static int _make_cursor(size_t index, size_t source, size_t len,
+                        mol2_source_t read, mol2_cursor_t *cur) {
+  ASSERT(cur);
+  ASSERT(len);
+  ASSERT(read);
+
+  int err = 0;
+
+  cur->offset = 0;
+  cur->size = len;
+
+  memset(g_mol_data_source, 0, sizeof(g_mol_data_source));
+  mol2_data_source_t *ptr = (mol2_data_source_t *)g_mol_data_source;
+
+  ptr->read = read;
+  ptr->total_size = len;
+
+  ptr->args[0] = index;
+  ptr->args[1] = source;
+
+  ptr->cache_size = 0;
+  ptr->start_point = 0;
+  ptr->max_cache_size = MAX_CACHE_SIZE;
+  cur->data_source = ptr;
+
+  return err;
+}
+
+static uint32_t _read_data_from_witness(uintptr_t arg[], uint8_t *ptr,
+                                        uint32_t len, uint32_t offset) {
+  int err;
+  uint64_t output_len = len;
+  err = ckb_load_witness(ptr, &output_len, offset, arg[0], arg[1]);
+  if (err != 0) {
+    return 0;
+  }
+  if (output_len > len) {
+    return len;
+  } else {
+    return (uint32_t)output_len;
+  }
+}
+
+int _get_cursor_from_witness(WitnessArgsType *witness, size_t index,
+                             size_t source) {
+  int err = CKB_SUCCESS;
+  mol2_cursor_t cur;
+  uint64_t len = 0;
+  ckb_load_witness(NULL, &len, 0, index, source);
+  if (len == 0) {
+    return ERROR_LOAD_WITNESS;
+  }
+
+  err = _make_cursor(index, source, len, _read_data_from_witness, &cur);
+  if (err != CKB_SUCCESS) {
+    return err;
+  }
+
+  *witness = make_WitnessArgs(&cur);
+  return err;
+}
 
 #ifdef CKB_USE_SIM
 int simulator_main() {
 #else
-int main(int argc, const char* argv[]) {
+int main(int argc, const char *argv[]) {
 #endif
-  return 0;
-}
+  // Get public key
+  uint8_t msg[32] = {0};
+  int error = generate_sighash_all(msg, sizeof(msg));
+  if (error != CKB_SUCCESS) {
+    return error;
+  }
+
+  size_t new_msg_len = generate_new_msg(NULL, 0, msg, sizeof(msg), NULL, 0);
+  if (new_msg_len == 0) {
+    return ERROR_GENERATE_NEW_MSG;
+  }
+  uint8_t new_msg[new_msg_len];
+  memset(new_msg, 0, new_msg_len);
+  if (generate_new_msg(new_msg, new_msg_len, msg, sizeof(msg), NULL, 0)) {
+    return ERROR_GENERATE_NEW_MSG;
+  }
```

Comment:
use a fixed value `new_msg_len` based on the size of `msg`, e.g. 100, 256. Don't calculate it on fly.